### PR TITLE
Fix softmax_warp_backward math when is_log_softmax = True and register LogSoftmax CUDA kernel

### PR DESCRIPTION
**Pull request by suffiank**
**Description**: 
- Register LogSoftmax CUDA kernel.
- Fix math for _softmax_warp_backward_ when is_log_softmax = true. 
- Fix operator tests to actually invoke _softmax_warp_forward_ and _softmax_warp_backward_ as well as CUDA DNN fallbacks. 
- Remove what seems like unnecessary constraint for axis to be on last dimension? Comment?

**Motivation and Context**
- The LogSoftmax CUDA kernel was not registered and execution was falling back to LogSoftmax CPU kernel.
- The gradient checker was testing LogSoftmaxGrad CPU kernel (not CUDA kernel).
- The _softmax_warp_forward_ and _softmax_warp_backward_ were not being tested.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/math/softmax.cc [hunk 1]
```diff
@@ -26,7 +26,7 @@ Status SoftMaxComputeHelper(
 
   // cudnnSoftmaxForward/Backward is not optimal implementation.
   // TODO: remove cudnn path completely in the future.
-  if (D == input_shape[normalized_axis] && D <= 1024 && D * sizeof(T) <= 4096) {
+  if (D <= 1024 && D * sizeof(T) <= 4096) {
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/math/softmax.cc:29]: I think we can relax the "D == input_shape[normalized_axis]" condition... 
But let's also add some test cases to cover this .

Comment (suffiank) [reply onnxruntime/core/providers/cuda/math/softmax.cc:29]: The existing tests are invoking this case (i.e. D != input_shape[normalized_axis]) since shape is {8, 2, 128} and an unspecified axis defaults to one. But I can add a different shape/axis combination.

File: orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu [hunk 2]
```diff
@@ -77,15 +77,28 @@ __global__ void softmax_warp_backward(output_t* gradInput, const input_t* grad,
   }
 
   acc_t sum[WARP_BATCH];
-#pragma unroll
-  for (int i = 0; i < WARP_BATCH; ++i) {
-    sum[i] = grad_output_reg[i][0];
-#pragma unroll
-    for (int it = 1; it < WARP_ITERATIONS; ++it) {
-      sum[i] += grad_output_reg[i][it];
+  if (!is_log_softmax) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu:80]: I am uncertain about this change... 

See 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/PersistentSoftmax.cuh#L156

Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu:80]: For pytorch's backward path, there is only one space where it need to switch backed on is_log_softmax

File: onnxruntime/core/providers/cuda/math/softmax.cc (hunk 1 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/math/softmax.cc:29]: good point... I though the default axis is -1 :)

File: orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu (hunk 2 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu:80]: So the is_log_softmax = True expressions are identical between the files. Note the definition of `sum[i]` differs in their case. This expression also matches the one derived earlier for the CPU kernel: `dX_i = d(log Y_i) - [sum_j d(log Y_j)] exp(log Y_i)` for element _i_. It's also not too hard to derive `dX_i = Y_i [ dY_i - sum_j Y_j dY_j  ]` as we use for SoftmaxGrad. Their expression differs by appearance for SoftmaxGrad case. I will think on it.

File: onnxruntime/core/providers/cuda/math/softmax.cc (hunk 1 above)
Comment (suffiank) [reply onnxruntime/core/providers/cuda/math/softmax.cc:29]: Add tests for all axes (considering last dimension as special case).

Comment (suffiank) [reply onnxruntime/core/providers/cuda/math/softmax.cc:29]: Oh, I missed your comment. Ha :)

File: orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu (hunk 2 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu:80]: I'm not making sense of the SoftmaxGrad expression from PersistentSoftmax.cuh to be honest.
Their expression is `dX_i = dY_i - Y_i sum_j dY_j`, right?

Consider the specific case `loss := Y_0 + Y_1`. 
Then `Y_0(X_0, X_1) + Y_1(X_0, X_1) = 1` and `dY_0 = dY_1 = 1` and `dX_0 = dX_1 = 0`, right?

That jives with `dX_i = Y_i [ dY_i - sum_j Y_j dY_j ]`: `dX_i = Y_i ( 1 - Y_0 - Y_1 ) = 0.` Check.
It does NOT with `dX_i = dY_i - Y_i sum_j dY_j:` `0 = 1 - 2 Y_i` implies `Y_i = 1/2.` False.

And our expression derives as (.. letting `Z := sum_i exp(X_i)` be the normalization factor..):
```
dX_i := d(loss)/dX_i 
      = sum_j d(loss)/dY_j dY_j/dX_i 
      = sum_j d(loss)/dY_j [ exp(X_i)/Z kronecker_delta_ij - exp(X_j)/Z^2 exp(X_i)]
      = sum_j d(loss)/dY_j [ Y_i kronecker_delta_ij - Y_i Y_j ]
      = d(loss)/dY_i Y_i  - sum_j d(loss)/dY_j Y_i Y_j
      = Y_i [ d(loss)/dY_i - sum_j d(loss)/dY_j Y_j ]
      = Y_i [ dY_i - sum_j dY_j Y_j ]
```

So I'm not sure what's going on in their case.

Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu:80]: Also see apex [softmax_warp_backward](https://github.com/NVIDIA/apex/blob/4ef930c1c884fdca5f472ab2ce7cb9b505d26c1a/apex/contrib/csrc/multihead_attn/softmax.h#L1606)..

**Review by SherlockNoMad — APPROVED** [review]
